File: backend/app/services/email_campaign_service.py
import os
import logging
from typing import List, Dict, Any, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_

from app.models import EmailTemplate, EmailCampaign, EmailRecipient, EmailList, Contact, Project
from app.schemas import (
    EmailTemplateCreate, EmailTemplateUpdate,
    EmailCampaignCreate, EmailCampaignUpdate,
    EmailRecipientCreate, EmailListCreate, EmailListUpdate
)

logger = logging.getLogger(__name__)


class EmailCampaignService:
    def __init__(self, db: Session):
        self.db = db
        self.sendgrid_api_key = os.getenv("SENDGRID_API_KEY")
        
    # Email Templates
    def create_template(self, template_data: EmailTemplateCreate, user_id: int) -> EmailTemplate:
        """Créer un nouveau template email"""
        template = EmailTemplate(
            name=template_data.name,
            type=template_data.type,
            subject=template_data.subject,
            body_html=template_data.body_html,
            variables=template_data.variables,
            created_by_user_id=user_id
        )
        self.db.add(template)
        self.db.commit()
        self.db.refresh(template)
        return template
    
    def get_templates(self, user_id: int, template_type: Optional[str] = None) -> List[EmailTemplate]:
        """Récupérer les templates email"""
        query = self.db.query(EmailTemplate).filter(EmailTemplate.created_by_user_id == user_id)
        if template_type:
            query = query.filter(EmailTemplate.type == template_type)
        return query.order_by(EmailTemplate.created_at.desc()).all()
    
    def get_template(self, template_id: int, user_id: int) -> Optional[EmailTemplate]:
        """Récupérer un template par ID"""
        return self.db.query(EmailTemplate).filter(
            and_(
                EmailTemplate.id == template_id,
                EmailTemplate.created_by_user_id == user_id
            )
        ).first()
    
    def update_template(self, template_id: int, template_data: EmailTemplateUpdate, user_id: int) -> Optional[EmailTemplate]:
        """Mettre à jour un template"""
        template = self.get_template(template_id, user_id)
        if not template:
            return None
        
        if template_data.name is not None:
            template.name = template_data.name
        if template_data.type is not None:
            template.type = template_data.type
        if template_data.subject is not None:
            template.subject = template_data.subject
        if template_data.body_html is not None:
            template.body_html = template_data.body_html
        if template_data.variables is not None:
            template.variables = template_data.variables
        
        template.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(template)
        return template
    
    def delete_template(self, template_id: int, user_id: int) -> bool:
        """Supprimer un template"""
        template = self.get_template(template_id, user_id)
        if not template:
            return False
        
        self.db.delete(template)
        self.db.commit()
        return True
    
    # Email Campaigns
    def create_campaign(self, campaign_data: EmailCampaignCreate, user_id: int) -> EmailCampaign:
        """Créer une nouvelle campagne email"""
        campaign = EmailCampaign(
            project_id=campaign_data.project_id,
            template_id=campaign_data.template_id,
            name=campaign_data.name,
            subject=campaign_data.subject,
            body_html=campaign_data.body_html,
            scheduled_at=campaign_data.scheduled_at,
            created_by_user_id=user_id
        )
        self.db.add(campaign)
        self.db.commit()
        self.db.refresh(campaign)
        return campaign
    
    def get_campaigns(self, user_id: int, project_id: Optional[int] = None) -> List[EmailCampaign]:
        """Récupérer les campagnes email"""
        query = self.db.query(EmailCampaign).filter(EmailCampaign.created_by_user_id == user_id)
        if project_id:
            query = query.filter(EmailCampaign.project_id == project_id)
        return query.order_by(EmailCampaign.created_at.desc()).all()
    
    def get_campaign(self, campaign_id: int, user_id: int) -> Optional[EmailCampaign]:
        """Récupérer une campagne par ID"""
        return self.db.query(EmailCampaign).filter(
            and_(
                EmailCampaign.id == campaign_id,
                EmailCampaign.created_by_user_id == user_id
            )
        ).first()
    
    def update_campaign(self, campaign_id: int, campaign_data: EmailCampaignUpdate, user_id: int) -> Optional[EmailCampaign]:
        """Mettre à jour une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign:
            return None
        
        if campaign_data.name is not None:
            campaign.name = campaign_data.name
        if campaign_data.subject is not None:
            campaign.subject = campaign_data.subject
        if campaign_data.body_html is not None:
            campaign.body_html = campaign_data.body_html
        if campaign_data.status is not None:
            campaign.status = campaign_data.status
        if campaign_data.scheduled_at is not None:
            campaign.scheduled_at = campaign_data.scheduled_at
        
        campaign.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(campaign)
        return campaign
    
    def delete_campaign(self, campaign_id: int, user_id: int) -> bool:
        """Supprimer une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign:
            return False
        
        self.db.delete(campaign)
        self.db.commit()
        return True
    
    # Email Recipients
    def add_recipients_to_campaign(self, campaign_id: int, recipients_data: List[EmailRecipientCreate], user_id: int) -> List[EmailRecipient]:
        """Ajouter des destinataires à une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign:
            return []
        
        recipients = []
        for recipient_data in recipients_data:
            recipient = EmailRecipient(
                campaign_id=campaign_id,
                contact_id=recipient_data.contact_id,
                email=recipient_data.email,
                name=recipient_data.name
            )
            recipients.append(recipient)
            self.db.add(recipient)
        
        # Mettre à jour le nombre total de destinataires
        campaign.total_recipients = len(self.db.query(EmailRecipient).filter(EmailRecipient.campaign_id == campaign_id).all())
        
        self.db.commit()
        for recipient in recipients:
            self.db.refresh(recipient)
        return recipients
    
    def get_campaign_recipients(self, campaign_id: int, user_id: int) -> List[EmailRecipient]:
        """Récupérer les destinataires d'une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign:
            return []
        
        return self.db.query(EmailRecipient).filter(EmailRecipient.campaign_id == campaign_id).all()
    
    def get_campaign_stats(self, campaign_id: int, user_id: int) -> Dict[str, Any]:
        """Récupérer les statistiques d'une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign:
            return {}
        
        recipients = self.get_campaign_recipients(campaign_id, user_id)
        
        stats = {
            "campaign_id": campaign_id,
            "total_recipients": len(recipients),
            "sent_count": len([r for r in recipients if r.status in ["sent", "opened", "clicked"]]),
            "opened_count": len([r for r in recipients if r.status in ["opened", "clicked"]]),
            "clicked_count": len([r for r in recipients if r.status == "clicked"]),
            "bounced_count": len([r for r in recipients if r.status == "bounced"]),
        }
        
        # Calculer les taux
        if stats["total_recipients"] > 0:
            stats["open_rate"] = stats["opened_count"] / stats["total_recipients"]
            stats["click_rate"] = stats["clicked_count"] / stats["total_recipients"]
            stats["bounce_rate"] = stats["bounced_count"] / stats["total_recipients"]
        else:
            stats["open_rate"] = 0.0
            stats["click_rate"] = 0.0
            stats["bounce_rate"] = 0.0
        
        return stats
    
    # Email Lists
    def create_email_list(self, list_data: EmailListCreate, user_id: int) -> EmailList:
        """Créer une nouvelle liste de diffusion"""
        email_list = EmailList(
            name=list_data.name,
            description=list_data.description,
            contact_ids=list_data.contact_ids,
            filters=list_data.filters,
            created_by_user_id=user_id
        )
        
        # Calculer le nombre total de contacts
        if list_data.contact_ids:
            email_list.total_contacts = len(list_data.contact_ids)
        elif list_data.filters:
            # Appliquer les filtres pour compter les contacts
            email_list.total_contacts = self._count_contacts_by_filters(list_data.filters, user_id)
        
        self.db.add(email_list)
        self.db.commit()
        self.db.refresh(email_list)
        return email_list
    
    def get_email_lists(self, user_id: int) -> List[EmailList]:
        """Récupérer les listes de diffusion"""
        return self.db.query(EmailList).filter(EmailList.created_by_user_id == user_id).order_by(EmailList.created_at.desc()).all()
    
    def get_email_list(self, list_id: int, user_id: int) -> Optional[EmailList]:
        """Récupérer une liste par ID"""
        return self.db.query(EmailList).filter(
            and_(
                EmailList.id == list_id,
                EmailList.created_by_user_id == user_id
            )
        ).first()
    
    def update_email_list(self, list_id: int, list_data: EmailListUpdate, user_id: int) -> Optional[EmailList]:
        """Mettre à jour une liste de diffusion"""
        email_list = self.get_email_list(list_id, user_id)
        if not email_list:
            return None
        
        if list_data.name is not None:
            email_list.name = list_data.name
        if list_data.description is not None:
            email_list.description = list_data.description
        if list_data.contact_ids is not None:
            email_list.contact_ids = list_data.contact_ids
            email_list.total_contacts = len(list_data.contact_ids)
        if list_data.filters is not None:
            email_list.filters = list_data.filters
            email_list.total_contacts = self._count_contacts_by_filters(list_data.filters, user_id)
        
        email_list.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(email_list)
        return email_list
    
    def delete_email_list(self, list_id: int, user_id: int) -> bool:
        """Supprimer une liste de diffusion"""
        email_list = self.get_email_list(list_id, user_id)
        if not email_list:
            return False
        
        self.db.delete(email_list)
        self.db.commit()
        return True
    
    def _count_contacts_by_filters(self, filters: Dict[str, Any], user_id: int) -> int:
        """Compter les contacts selon les filtres"""
        query = self.db.query(Contact).filter(Contact.user_id == user_id)
        
        if "company_id" in filters:
            query = query.filter(Contact.company_id == filters["company_id"])
        if "status" in filters:
            query = query.filter(Contact.status == filters["status"])
        if "tags" in filters:
            # Recherche dans les tags JSON
            for tag in filters["tags"]:
                query = query.filter(Contact.tags.contains([tag]))
        
        return query.count()
    
    def get_contacts_from_list(self, list_id: int, user_id: int) -> List[Contact]:
        """Récupérer les contacts d'une liste de diffusion"""
        email_list = self.get_email_list(list_id, user_id)
        if not email_list:
            return []
        
        if email_list.contact_ids:
            # Liste statique d'IDs
            return self.db.query(Contact).filter(
                and_(
                    Contact.id.in_(email_list.contact_ids),
                    Contact.user_id == user_id
                )
            ).all()
        elif email_list.filters:
            # Liste dynamique basée sur les filtres
            query = self.db.query(Contact).filter(Contact.user_id == user_id)
            
            if "company_id" in email_list.filters:
                query = query.filter(Contact.company_id == email_list.filters["company_id"])
            if "status" in email_list.filters:
                query = query.filter(Contact.status == email_list.filters["status"])
            if "tags" in email_list.filters:
                for tag in email_list.filters["tags"]:
                    query = query.filter(Contact.tags.contains([tag]))
            
            return query.all()
        
        return []
    
    # SendGrid Integration (simulation pour le développement)
    def send_campaign(self, campaign_id: int, user_id: int) -> bool:
        """Envoyer une campagne email (simulation)"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign or campaign.status != "draft":
            return False
        
        # Simulation d'envoi
        campaign.status = "sending"
        campaign.sent_at = datetime.utcnow()
        
        recipients = self.get_campaign_recipients(campaign_id, user_id)
        
        # Simuler l'envoi à chaque destinataire
        for recipient in recipients:
            recipient.status = "sent"
            recipient.sent_at = datetime.utcnow()
        
        campaign.status = "sent"
        campaign.sent_count = len(recipients)
        
        self.db.commit()
        
        logger.info(f"Campaign {campaign_id} sent to {len(recipients)} recipients")
        return True
    
    def schedule_campaign(self, campaign_id: int, scheduled_at: datetime, user_id: int) -> bool:
        """Planifier l'envoi d'une campagne"""
        campaign = self.get_campaign(campaign_id, user_id)
        if not campaign or campaign.status != "draft":
            return False
        
        campaign.status = "scheduled"
        campaign.scheduled_at = scheduled_at
        
        self.db.commit()
        return True

File: domains/informatique/02-projects/portfolio/Portfolio/angular-portfolio/src/app/core/services/email-campaign.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface EmailTemplate {
  id: number;
  name: string;
  type: string;
  subject: string;
  body_html: string;
  variables?: any;
  created_at: string;
  updated_at: string;
  created_by_user_id: number;
}

export interface EmailCampaign {
  id: number;
  project_id?: number;
  template_id: number;
  name: string;
  subject: string;
  body_html: string;
  status: string;
  scheduled_at?: string;
  sent_at?: string;
  total_recipients: number;
  sent_count: number;
  opened_count: number;
  clicked_count: number;
  bounced_count: number;
  created_at: string;
  updated_at: string;
  created_by_user_id: number;
}

export interface EmailRecipient {
  id: number;
  campaign_id: number;
  contact_id?: number;
  email: string;
  name?: string;
  status: string;
  sent_at?: string;
  opened_at?: string;
  clicked_at?: string;
  bounced_at?: string;
  open_count: number;
  click_count: number;
  created_at: string;
}

export interface EmailList {
  id: number;
  name: string;
  description?: string;
  contact_ids?: number[];
  filters?: any;
  total_contacts: number;
  created_at: string;
  updated_at: string;
  created_by_user_id: number;
}

export interface CampaignStats {
  campaign_id: number;
  total_recipients: number;
  sent_count: number;
  opened_count: number;
  clicked_count: number;
  bounced_count: number;
  open_rate: number;
  click_rate: number;
  bounce_rate: number;
}

export interface EmailDashboardStats {
  total_campaigns: number;
  total_templates: number;
  total_lists: number;
  campaigns_by_status: any;
  recent_campaigns: number;
  recent_campaigns_data: any[];
}

@Injectable({
  providedIn: 'root'
})
export class EmailCampaignService {
  private apiUrl = `${environment.apiUrl}/email-campaigns`;
  private dashboardStatsSubject = new BehaviorSubject<EmailDashboardStats | null>(null);
  public dashboardStats$ = this.dashboardStatsSubject.asObservable();

  constructor(private http: HttpClient) {}

  // Email Templates
  createTemplate(template: Partial<EmailTemplate>): Observable<EmailTemplate> {
    return this.http.post<EmailTemplate>(`${this.apiUrl}/templates`, template);
  }

  getTemplates(templateType?: string): Observable<EmailTemplate[]> {
    const params = templateType ? { template_type: templateType } : {};
    return this.http.get<EmailTemplate[]>(`${this.apiUrl}/templates`, { params: params as any });
  }

  getTemplate(templateId: number): Observable<EmailTemplate> {
    return this.http.get<EmailTemplate>(`${this.apiUrl}/templates/${templateId}`);
  }

  updateTemplate(templateId: number, template: Partial<EmailTemplate>): Observable<EmailTemplate> {
    return this.http.put<EmailTemplate>(`${this.apiUrl}/templates/${templateId}`, template);
  }

  deleteTemplate(templateId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/templates/${templateId}`);
  }

  // Email Campaigns
  createCampaign(campaign: Partial<EmailCampaign>): Observable<EmailCampaign> {
    return this.http.post<EmailCampaign>(`${this.apiUrl}/campaigns`, campaign);
  }

  getCampaigns(projectId?: number): Observable<EmailCampaign[]> {
    const params = projectId ? { project_id: projectId } : {};
    return this.http.get<EmailCampaign[]>(`${this.apiUrl}/campaigns`, { params: params as any });
  }

  getCampaign(campaignId: number): Observable<EmailCampaign> {
    return this.http.get<EmailCampaign>(`${this.apiUrl}/campaigns/${campaignId}`);
  }

  updateCampaign(campaignId: number, campaign: Partial<EmailCampaign>): Observable<EmailCampaign> {
    return this.http.put<EmailCampaign>(`${this.apiUrl}/campaigns/${campaignId}`, campaign);
  }

  deleteCampaign(campaignId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/campaigns/${campaignId}`);
  }

  sendCampaign(campaignId: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/campaigns/${campaignId}/send`, {});
  }

  scheduleCampaign(campaignId: number, scheduledAt: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/campaigns/${campaignId}/schedule`, { scheduled_at: scheduledAt });
  }

  getCampaignStats(campaignId: number): Observable<CampaignStats> {
    return this.http.get<CampaignStats>(`${this.apiUrl}/campaigns/${campaignId}/stats`);
  }

  // Email Recipients
  addRecipientsToCampaign(campaignId: number, recipients: Partial<EmailRecipient>[]): Observable<EmailRecipient[]> {
    return this.http.post<EmailRecipient[]>(`${this.apiUrl}/campaigns/${campaignId}/recipients`, recipients);
  }

  getCampaignRecipients(campaignId: number): Observable<EmailRecipient[]> {
    return this.http.get<EmailRecipient[]>(`${this.apiUrl}/campaigns/${campaignId}/recipients`);
  }

  // Email Lists
  createEmailList(emailList: Partial<EmailList>): Observable<EmailList> {
    return this.http.post<EmailList>(`${this.apiUrl}/lists`, emailList);
  }

  getEmailLists(): Observable<EmailList[]> {
    return this.http.get<EmailList[]>(`${this.apiUrl}/lists`);
  }

  getEmailList(listId: number): Observable<EmailList> {
    return this.http.get<EmailList>(`${this.apiUrl}/lists/${listId}`);
  }

  updateEmailList(listId: number, emailList: Partial<EmailList>): Observable<EmailList> {
    return this.http.put<EmailList>(`${this.apiUrl}/lists/${listId}`, emailList);
  }

  deleteEmailList(listId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/lists/${listId}`);
  }

  // Dashboard Stats
  getDashboardStats(): Observable<EmailDashboardStats> {
    return this.http.get<EmailDashboardStats>(`${this.apiUrl}/dashboard/stats`);
  }

  updateDashboardStats(): void {
    this.getDashboardStats().subscribe(stats => {
      this.dashboardStatsSubject.next(stats);
    });
  }

  // Helper methods
  getStatusColor(status: string): string {
    const statusColors: { [key: string]: string } = {
      'draft': '#6b7280',
      'scheduled': '#f59e0b',
      'sending': '#3b82f6',
      'sent': '#10b981',
      'failed': '#ef4444'
    };
    return statusColors[status] || '#6b7280';
  }

  getStatusLabel(status: string): string {
    const statusLabels: { [key: string]: string } = {
      'draft': 'Brouillon',
      'scheduled': 'Planifiée',
      'sending': 'En cours',
      'sent': 'Envoyée',
      'failed': 'Échec'
    };
    return statusLabels[status] || status;
  }

  formatDate(dateString: string): string {
    return new Date(dateString).toLocaleDateString('fr-FR', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    });
  }

  calculateOpenRate(campaign: EmailCampaign): number {
    if (campaign.total_recipients === 0) return 0;
    return (campaign.opened_count / campaign.total_recipients) * 100;
  }

  calculateClickRate(campaign: EmailCampaign): number {
    if (campaign.total_recipients === 0) return 0;
    return (campaign.clicked_count / campaign.total_recipients) * 100;
  }

  calculateBounceRate(campaign: EmailCampaign): number {
    if (campaign.total_recipients === 0) return 0;
    return (campaign.bounced_count / campaign.total_recipients) * 100;
  }
}
